src/core/coap/coap.cpp

```diff
@@ -209,6 +209,9 @@ Error CoapBase::SendMessage(Message                &aMessage,
     bool     moreBlocks           = false;
 #endif
 
+    // fire and forget (mAckTimeout=0) is only allowed for Non-confirmable (NON) messages
+    VerifyOrExit(aTxParameters.mAckTimeout != 0 || aMessage.IsNonConfirmable(), error = OT_ERROR_INVALID_ARGS);
```

Comment:
I suggest changing this. While this check is logically correct, using a negative boolean check with `||` can be harder to read and follow.

```c++
    // fire and forget (mAckTimeout=0) is only allowed for Non-confirmable (NON) messages
    VerifyOrExit(aTxParameters.mAckTimeout != 0 || aMessage.IsNonConfirmable(), error = OT_ERROR_INVALID_ARGS);
```

I would change it to the following, which reads better:

```c++
    if (aTxParameters.mAckTimeout == 0)
    {
        VerifyOrExit(aMessage.IsNonConfirmable(), error = kErrorInvalidArgs);
    }
```

Also, `kErrorInvalidArgs` should be used here instead of `OT_ERROR_INVALID_ARGS`.